New field in links (method, headers, body) (#1253)
radiantearth/stac-spec: item-spec/common-metadata.md

| href       | string | **REQUIRED.** The actual link in the format of an URL. Relative and absolute links are both allowed. [Trailing slashes are significant.](../best-practices.md#consistent-uris) |
| rel        | string | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |
| type       | string | [Media type](../catalog-spec/catalog-spec.md#media-types) of the referenced entity.                                                                                            |
| title      | string | A human readable title to be used in rendered displays of the link.                                                                                                            |
| method     | string | The HTTP method required to be supported by the target resource. `GET` by default                                                                                              |

> | headers    | object | The HTTP headers required to be sent by the target resource.                                                                                                                   |

Reviewer: We don't use the object type in the spec, instead we use `Map<string, *>` (applies to headers and body for now)

Reviewer: `Map<string, string|array<string>>`

`{a: "b"}`
translates to:
`a: b`

`{a: ["b", "c"]}`
translates to:
```
a: b
a: c
```

This could hopefully also be clarified in the STAC API spec.

This object describes a relationship with another entity. Data providers are advised to be liberal
with the links section, to describe things like the Catalog an Item is in, related Items, parent or
child Items (modeled in different ways, like an 'acquisition' or derived data).
It is allowed to add additional fields such as a `title` and `type`.

| Field Name | Type           | Description                                                                                                                                                                    |
| ---------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| href       | string         | **REQUIRED.** The actual link in the format of an URL. Relative and absolute links are both allowed. [Trailing slashes are significant.](../best-practices.md#consistent-uris) |
| rel        | string         | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |

> | type       | string         | [Media type](../catalog-spec/catalog-spec.md#media-types) of the referenced entity.                                                                                            |

Reviewer: I guess this shouldn't link to the catalog-spec. 

```suggestion
| type       | string         | Media type) of the referenced entity.                                                                                            |
```

| ---------- | -------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| href       | string                           | **REQUIRED.** The actual link in the format of an URL. Relative and absolute links are both allowed. [Trailing slashes are significant.](../best-practices.md#consistent-uris) |
| rel        | string                           | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |
| type       | string                           | Media type of the referenced entity.                                                                                                                                           |
| title      | string                           | A human readable title to be used in rendered displays of the link.                                                                                                            |

> | method     | string                           | The HTTP method that shall be used for the request to the target resource. `GET` by default                                                                                    |

Reviewer: ```suggestion
| method     | string                           | The HTTP method that shall be used for the request to the target resource, in uppercase. `GET` by default                                                                      |
```

PlanetScope Ortho Tile Product has an `gsd` of 3.7 (or 4 if rounding), even though the pixel size of the images is 3.125.
For example, one might choose for WorldView-2 the Multispectral 20° off-nadir value of 2.07
and for WorldView-3 the Multispectral 20° off-nadir value of 1.38.

### Link Object

This object describes a relationship with another entity. Data providers are advised to be liberal
with the links section, to describe things like the Catalog an Item is in, related Items, parent or
child Items (modeled in different ways, like an 'acquisition' or derived data).

> It is allowed to add additional fields such as a `title` and `type`.

Reviewer 2: Is this sentence correct? `title` and `type` are both fields described in the table below.

| Field Name | Type                             | Description                                                                                                                                                                    |
| ---------- | -------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| href       | string                           | **REQUIRED.** The actual link in the format of an URL. Relative and absolute links are both allowed. [Trailing slashes are significant.](../best-practices.md#consistent-uris) |
| rel        | string                           | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |

> | type       | string                           | [Media type](../catalog-spec/catalog-spec.md#media-types) of the referenced entity.                                                                                            |

Reviewer: ```suggestion
| type       | string                           | Media type of the referenced entity.                                                                                            |
```

| title      | string                           | A human readable title to be used in rendered displays of the link.                                                                                                            |

> | method     | string                           | The HTTP method required to be supported by the target resource. `GET` by default                                                                                              |

Reviewer: ```suggestion
| method     | string                           | The HTTP method that shall be used for the request to the target resource. `GET` by default                                                                                              |
```

> | headers    | Map<string, string \| \[string]> | The HTTP headers required to be sent by the target resource.                                                                                                                   |

Reviewer: ```suggestion
| headers    | Map<string, string \| \[string]> | The HTTP headers to be sent for the request to the target resource.                                                                                                                    |
```

> | body       | any                              | The HTTP body required to be sent by the target resource.                                                                                                                      |

Reviewer: ```suggestion
| body       | any                              | The HTTP body to be sent to the target resource.                                                                                                                      |
```

#### HTTP headers

The field `headers` allows to describe a dictionary of HTTP headers that are required to be sent by the client.
The keys of the dictionary are the header names, and the values are either a single string or an array of strings.
In case of an array, the header is expected to be sent multiple times with the different values.

For a full discussion of the situations where relative and absolute links are recommended see the

> ['Use of links'](../best-practices.md#use-of-links) section of the STAC best practices.

Reviewer: This seems unrelated to HTTP headers. Should this be in the other section below the table?

| href       | string         | **REQUIRED.** The actual link in the format of an URL. Relative and absolute links are both allowed. [Trailing slashes are significant.](../best-practices.md#consistent-uris) |
| rel        | string         | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |
| type       | string         | [Media type](../catalog-spec/catalog-spec.md#media-types) of the referenced entity.                                                                                            |
| title      | string         | A human readable title to be used in rendered displays of the link.                                                                                                            |
| method     | string         | The HTTP method required to be supported by the target resource. `GET` by default                                                                                              |

> | headers    | Map<string, *> | The HTTP headers required to be sent by the target resource.                                                                                                                   |

Reviewer: I think we still need to explain what we discussed in the last call. Arrays will be expanded to multiple headers.

Ultimately, we could also say:
`Map<string, string | [string]>`

Author: explaination added

Reviewer: Can't find the explanation. Where can it be found?

> | body       | object         | The HTTP body required to be sent by the target resource.                                                                                                                      |

Reviewer: ```suggestion
| body       | Map<string, *> | The HTTP body required to be sent by the target resource.                                                                                                                      |
```

Author: why the body as Map? body can be anything like the [stac api spec](https://github.com/radiantearth/stac-api-spec/blob/604ade6158de15b8ab068320ca41e25e2bf0e116/core/commons.yaml#L421), no?

Reviewer: Right, you are correct. But then it's also not an object. So it should be:

```suggestion
| body       | any            | The HTTP body required to be sent by the target resource.                                                                                                                      |
```
